joomla/joomla-cms
# [5.0] Media Manager Image crop using importmaps

Pull Request for Issue # .

### Summary of Changes
- Expose/localise the ESM module of the Cropper
- Deprecate the Global Cropper script
- Add the cropper.esm.js (the module) to the Joomla.assets.json (WAM) as importmap
- Use the natural way to import the module `import Cropper from 'cropper-module';` in the crop script


### Testing Instructions

- Check out this PR
- run `rpm ci`
- Open the media manager, select an image and edit it. Check that the CROP is working as expected

### Actual result BEFORE applying this Pull Request

Cropper is a Global Object

### Expected result AFTER applying this Pull Request

Cropper is ONLY available in the scope of the crop.js!

#### How importmap works:

There 2 parts: the import statement in the file that wants to import another script (ESM module) and the exposure of the module to the WAM (check the files in this PR). In the client side what happens is:

- A polyfill is loaded so browsers that don't support importmap won't crash:
<img width="1721" alt="Screenshot 2023-06-24 at 11 54 33" src="https://github.com/joomla/joomla-cms/assets/3889375/9e90034f-5656-40c9-a75c-d736c6170367">

- The polyfill does a simple check by trying to import a blob and either patch or not the browser:
<img width="1710" alt="Screenshot 2023-06-24 at 11 54 45" src="https://github.com/joomla/joomla-cms/assets/3889375/36e4405d-55a0-418f-9868-b58db83914fb">

- The file that wants to import the module then is parsed/executed:
<img width="1710" alt="Screenshot 2023-06-24 at 11 55 00" src="https://github.com/joomla/joomla-cms/assets/3889375/156c13b6-a16c-4afa-a052-e4d8b01f342d">

- The resolution of the module is taken from the importmap:
<img width="1102" alt="Screenshot 2023-06-24 at 12 12 50" src="https://github.com/joomla/joomla-cms/assets/3889375/8125e776-3342-46e6-bea2-a0cb4085220d">

- Finally the module is fetched/parsed/executed and the importer script continues:
<img width="1709" alt="Screenshot 2023-06-24 at 11 55 13" src="https://github.com/joomla/joomla-cms/assets/3889375/be23ad35-f9ac-47f2-8c4d-cfcd100c3e5f">


### Link to documentations
Please select:
